cantools/db/database.py

```diff
@@ -24,6 +24,7 @@ def __init__(self,
                  nodes=None,
                  buses=None,
                  version=None,
+                 attributes=None,
```

Comment:
What's the reason for adding a list attributes to the database? I mean, isn't this information already part of messages, signals and nodes? Am I missing something?

Reply:
I don't have access to the Vector tools myself, could you take a few relevant screenshots to give me a better understanding of the database structure?

The main goal for this package is to make it easy to use (mainly to encode and decode messages), and at the same time complete. The second most important goal is to make the interface independent of the database format. When it comes to attributes the goal should be to make them useful to the user. I think a dictionary of attribute definitions (BA_DEF_) with the attribute name as key and various data (range, type, ...) as value may be useful. The attribute definition default (BA_DEF_DEF_) should be part of this value, and not be in a separate dictionary as today.

When it comes to signal, message and node attributes, they are probably most often accessed from the signal, message or node objects. `send_type` and `cycle_time` are two examples of message attributes. Maybe a separate attribute list could be useful, but I'm not sure. Is there a use case?

Reply:
You are right in that my perspective was very much from the dbc side of things. I just thought that serialisation (dump) of attributes would be a mess again if I scatter them all over the place. But from a usage perspective 1.) iterating the aggregate is ugly when one could use dicts and 2.) navigating multiple collections to harvest all info on one attribute is too.

I think I will come up with a more usage friendly approach. If you have further wishes that I shoud consider when implementing, do not hesitate.

My use case BTW is just dbc IO. I read them in, manipulate them, and dump them out. Since I use the dbcs with CANoe Interaction Layers in test setups, I need support for all these colorful attributes that control the layers behaviour.

A second use case for me is to render html doc from dbcs.

Reply:
I would be lovely if you can write an example showcasing your use case. Could help others as well.

Other than that I think my previous comment summarizes my current view of how to implement it. Though, I'm open to have internal representations of the attributes if it simplifies the implementation and makes the code easier to understand and maintain.

At last, most importantly, user friendliness is key to success!

Reply:
Also, all members prefixed with and underscore should not be considered private and should not be used directly by the user. So I guess a we should create a property allowing the user access to attribute definitions, and other useful data if any.

Reply:
Hi there,

I will proceed as follows: Make an independent attribute definition dict (by name). Database will hold that. The attributes themselves will be objects containing value and a reference to the respective definition. The attributes will be stored in dicts in node/signal/message.

I think that was your proposal, correct?

Reply:
It's very hard to design the message/signal/node-classes to be database format agnostic. Preferably DBC-naming of certain attributes should not be part of these classes, today `send_type`, etc. is the abstraction in these classes. I realize that I don't know enough about other database formats to make a good design of cantools package, so we have to implement something based on what we know now and what we need. I think we should proceed as outlined in your comment. Maybe think once more about the attribute dictionary in signals/messages/nodes. At least the new implementation should not clobber other database formats implementations.

Reply:
Ok. I am on it right now. To avoid redundant storage, i want to change send_type into properties that lookup in the dicts on the fly.

Since it would be ugly to have a back reference to the top level (database) to access the definitions, I propose to give node/message/signal each their own reference to the definition dict. That solves the issue of fallback to default when an attribute is not available. Do you agree?

Reply:
Lets discuss that later. I will stick to the current implementation right now. Doing lookups is complicating interoperability with other formats.

Reply:
Please create a new PR with the minor fixes (standard id, whitespace and valid node names), and I'll merge that before the rest is ready.

Reply:
I don' know it it makes any sense, but what do you think about this:

https://github.com/eerimoq/cantools/blob/dbc-attributes/examples/dbc_builder.py

Since I have absolutely no idea about what different formats have in common I figured, why not just store all format specific data in a separate database attribute; for DBC the dbc attribute. I also kept attributes as lists just as you, and added a get method. Nodes and messages are already using the same pattern.

Reply:
It is a neat idea to keep format specifics in such a container. It is unlikely that other formats use exactly the same properties on attributes, so why litter the generic message with specifics. And one can still implement transformation functions if one need to translate attributes when changing formats.

I am half way there and already close to your proposal.

Do you insist on the intermediate PR? Would be a lot of extra effort for me. I promise I will up my granularity in the future :).

Reply:
Nah, I thought you paused the implementation and I was afraid I would lose valuable fixes.

Reply:
In your proposal, is the „dbc“ one shared object or is it just a „namespace“ so separate e.g. the signal-specific attribute collection?

If it is kind of a namespace, do I use a local class for that?

Reply:
Hi. Short update:

I had started using dicts for the node/message/signal attributes and also for the defaults. That deviates from your proposal but eases access by name. I have switched to lists briefly but hesitatet because I would need traversal functions (e.g. get_attribute_by_name) in several places.

Do you think you could live with that?

I have a working version now that still lacks the dbc specifics separation (that is message.attributes instead of message.dbc.attributes). If you could give me a hint how to do this separation, I would also implement that (Its just a few weeks of Python for me so far :).

Reply:
I have no string feeling about lists vs dicts at the moment. It could always be changed later. Most importantly the user interface should be easy to use.

This is one alternative:

``` python
class Database(object):

    @property
    def dbc(self):
        # Return a DBC object, probably called self._dbc.
```